# **Desafio 5 Martin Gomez**

# 1.Carga los datos y crea un DataFrame con cada uno de ellos.

In [31]:
#importar librerias necesarias para hacer el desafio
import pandas as pd
import numpy as np

df_officers = pd.read_pickle('/content/drive/MyDrive/Cosas importantes/Curso Desafio/Python/Curso de python/Apoyodesafio/officers.pkl')
df_incidents = pd.read_pickle('/content/drive/MyDrive/Cosas importantes/Curso Desafio/Python/Curso de python/Apoyodesafio/incidents.pkl')
df_subjects = pd.read_pickle('/content/drive/MyDrive/Cosas importantes/Curso Desafio/Python/Curso de python/Apoyodesafio/subjects.pkl')


In [ ]:
df_officers.head()

case_number race gender  last_name first_name         full_name
0      44523A    L      M     Patino    Michael   Patino, Michael
1      44523A    W      M  Fillingim      Brian  Fillingim, Brian
2     121982X    L      M    Padilla    Gilbert  Padilla, Gilbert
3     605484T    W      M     Poston      Jerry     Poston, Jerry
4     384832T    B      M      Mondy    Michael    Mondy, Michael

In [ ]:
df_subjects.head()

case_number race gender last_name first_name        full_name
0      44523A    L      M     Curry      James     Curry, James
1     121982X    L      M    Chavez    Gabriel  Chavez, Gabriel
2     605484T    L      M   Salinas       Nick    Salinas, Nick
3     384832T    B      M     Smith      James     Smith, James
4     384832T    B      M      Dews    Antonio    Dews, Antonio

# 2.Genera una tabla que contenga la unión de las 3 tablas

In [3]:
df = df_incidents.merge(df_officers, how="left", on="case_number").merge(df_subjects, how="left", on="case_number", suffixes=('_officer', 'subject'))

df.head()

case_number        date                  location subject_statuses  \
0      44523A  2013-02-23     3000 Chihuahua Street          Injured   
1      44523A  2013-02-23     3000 Chihuahua Street          Injured   
2     121982X  2010-05-03  1300 N. Munger Boulevard          Injured   
3     605484T  2007-08-12   200 S. Stemmons Freeway            Other   
4     384832T  2007-05-26           7900 S. Loop 12   Shoot and Miss   

  subject_weapon                                           subjects  \
0        Handgun                                   Curry, James L/M   
1        Handgun                                   Curry, James L/M   
2        Handgun                                Chavez, Gabriel L/M   
3        Shotgun                                  Salinas, Nick L/M   
4        Unarmed  Smith, James B/M; Dews, Antonio B/M; Spearman,...   

   subject_count                                   officers  officer_count  \
0              1  Patino, Michael L/M; Fillingim, Brian W/M              2   
1              1  Patino, Michael L/M; Fillingim, Brian W/M              2   
2              1                       Padilla, Gilbert L/M              1   
3              1                          Poston, Jerry W/M              1   
4              3                         Mondy, Michael B/M              1   

  grand_jury_disposition  ... race_officer gender_officer last_name_officer  \
0                No Bill  ...            L              M            Patino   
1                No Bill  ...            W              M         Fillingim   
2                No Bill  ...            L              M           Padilla   
3            See Summary  ...            W              M            Poston   
4                   None  ...            B              M             Mondy   

   first_name_officer  full_name_officer racesubject gendersubject  \
0             Michael    Patino, Michael           L             M   
1               Brian   Fillingim, Brian           L             M   
2             Gilbert   Padilla, Gilbert           L             M   
3               Jerry      Poston, Jerry           L             M   
4             Michael     Mondy, Michael           B             M   

  last_namesubject first_namesubject full_namesubject  
0            Curry             James     Curry, James  
1            Curry             James     Curry, James  
2           Chavez           Gabriel  Chavez, Gabriel  
3          Salinas              Nick    Salinas, Nick  
4            Smith             James     Smith, James  

[5 rows x 25 columns]

#  3.Verifica si hay filas duplicadas; si es así, elimínalas.

In [6]:
print(len(df_incidents))
print(len(df))

219
380


In [7]:
df.drop_duplicates(inplace=True)

# 4.¿Cuántos sujetos de género F hay en el DataFrame resultante?

In [12]:
# Contar sujetos de género F
Total_mujeres = df['gendersubject'].value_counts().get('F', 0)
print(f"Numero de mujeres totales: {Total_mujeres}")



Numero de mujeres totales: 9


# 5.¿Encuántos números de caso hay por lo menos una sospechosa que sea mujer?

In [16]:
# Filtrar casos con al menos una sospechosa mujer
Casos_femeninos = df[df['gendersubject'] == 'F']['case_number'].unique()
num_Casos_femeninos = len(Casos_femeninos)
print(f"números de caso hay por lo menos una sospechosa que sea mujer: {num_Casos_femeninos }")


números de caso hay por lo menos una sospechosa que sea mujer: 7


## 6.Genera una tabla pivote que muestre en las filas el género del oficial y en las columnas el género del subject. ¿Cómo interpretas los valores que muestra esta vista?

In [20]:
# Crear tabla pivote con el género del oficial y del sujeto
pivot_table = pd.pivot_table(df, values='case_number', index='gender_officer', columns='gendersubject', aggfunc='count')

# Mostrar la tabla pivote
pivot_table



gendersubject   F    M
gender_officer        
F               2   18
M               7  350


## Interpretacion
*   La mayoría de los casos involucran oficiales y sujetos masculinos (350 casos).
*   Hay significativamente menos casos con oficiales femeninas.
* Los casos con oficiales femeninas y sujetos femeninos son los menos frecuentes(2 casos).




#**Parte 2**


## 1.Para continuar con el desarrollo de este desafío, necesitarás el archivos Cleaned_DS_Jobs.csv

## 2.Carga los datos y crea un DataFrame con ellos.

In [22]:
#Cargar el csv
df = pd.read_csv("/content/drive/MyDrive/Cosas importantes/Curso Desafio/Python/Curso de python/Apoyodesafio/Cleaned_DS_Jobs.csv")
df.head()

Job Title Salary Estimate  \
0  Sr Data Scientist        137-171    
1     Data Scientist        137-171    
2     Data Scientist        137-171    
3     Data Scientist        137-171    
4     Data Scientist        137-171    

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

         Company Name       Location            Headquarters  \
0         Healthfirst   New York, NY            New York, NY   
1             ManTech  Chantilly, VA             Herndon, VA   
2      Analysis Group     Boston, MA              Boston, MA   
3             INFICON     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions   New York, NY            New York, NY   

                      Size        Type of ownership  \
0   1001 to 5000 employees   Nonprofit Organization   
1  5001 to 10000 employees         Company - Public   
2   1001 to 5000 employees  Private Practice / Firm   
3    501 to 1000 employees         Company - Public   
4      51 to 200 employees        Company - Private   

                                Industry  ... company_age python  excel  \
0                     Insurance Carriers  ...          27      0      0   
1                 Research & Development  ...          52      0      0   
2                             Consulting  ...          39      1      1   
3  Electrical & Electronic Manufacturing  ...          20      1      1   
4                Advertising & Marketing  ...          22      1      1   

   hadoop  spark aws  tableau  big_data        job_simp  seniority  
0       0      0   1        0         0  data scientist     senior  
1       1      0   0        0         1  data scientist         na  
2       0      0   1        0         0  data scientist         na  
3       0      0   1        0         0  data scientist         na  
4       0      0   0        0         0  data scientist         na  

[5 rows x 27 columns]

### 3. Utiliza la siguiente lista de valores que serán considerados como nulos: ["na", "NA",-1, "0", "-1", "null", "n/a", "N/A", "NULL"]

In [23]:
# Calcular la cantidad de valores nulos en cada columna del DataFrame
df.isnull().sum()

Job Title            0
Salary Estimate      0
Job Description      0
Rating               0
Company Name         0
Location             0
Headquarters         0
Size                 0
Type of ownership    0
Industry             0
Sector               0
Revenue              0
min_salary           0
max_salary           0
avg_salary           0
job_state            0
same_state           0
company_age          0
python               0
excel                0
hadoop               0
spark                0
aws                  0
tableau              0
big_data             0
job_simp             0
seniority            0
dtype: int64

In [27]:
#identicar el numero de valores nulos
valores_nulos = ["na", "NA", -1, "0", "-1", "null", "n/a", "N/A", "NULL"]

df.isin(valores_nulos).sum()

Job Title              0
Salary Estimate        0
Job Description        0
Rating                 0
Company Name           0
Location               0
Headquarters          31
Size                  27
Type of ownership     27
Industry              71
Sector                71
Revenue               27
min_salary             0
max_salary             0
avg_salary             0
job_state              0
same_state             0
company_age          118
python                 0
excel                  0
hadoop                 0
spark                  0
aws                    0
tableau                0
big_data               0
job_simp              68
seniority            567
dtype: int64

In [32]:
# Reemplazar los valores considerados nulos con np.nan
df.replace(valores_nulos, np.nan, inplace=True)
df.isnull().sum()

Job Title              0
Salary Estimate        0
Job Description        0
Rating                 0
Company Name           0
Location               0
Headquarters          31
Size                  27
Type of ownership     27
Industry              71
Sector                71
Revenue               27
min_salary             0
max_salary             0
avg_salary             0
job_state              0
same_state             0
company_age          118
python                 0
excel                  0
hadoop                 0
spark                  0
aws                    0
tableau                0
big_data               0
job_simp              68
seniority            567
dtype: int64

## 4.Elimina todas las filas con datos faltantes.

In [33]:
# Eliminar filas con datos faltantes
df.dropna(inplace=True)

## 5.Apartir de la columna “Salary Estimate”, genera dos columnas: Salario Estimado Mínimo y Máximo.

In [34]:
# Ver los salarios estimados que estan en el documento
df[['Salary Estimate']].head()

Salary Estimate
0         137-171 
32         75-131 
38         75-131 
45         75-131 
54         75-131

In [40]:
df['Min Estimated Salary'] = df['Salary Estimate'].apply(lambda x: x.split('-')[0])
df['Max Estimated Salary'] = df['Salary Estimate'].apply(lambda x: x.split('-')[1])

# Convertir las columnas de salario a números, manejando posibles errores para poder hacer la tabla pivote
df['Min Estimated Salary'] = pd.to_numeric(df['Min Estimated Salary'], errors='coerce')
df['Max Estimated Salary'] = pd.to_numeric(df['Max Estimated Salary'], errors='coerce')

df[['Salary Estimate','Min Estimated Salary','Max Estimated Salary']].head()

Salary Estimate  Min Estimated Salary  Max Estimated Salary
0         137-171                    137                   171
32         75-131                     75                   131
38         75-131                     75                   131
45         75-131                     75                   131
54         75-131                     75                   131

## *6*.Realiza la recodificación de la columna Size con los valores de la siguiente tabla:

In [36]:
# Recodificación de la columna Size
size_mapping = {
    "10000+ employees": "Mega Empresas",
    "5001 to 10000 employees": "Grandes Empresas",
    "1001 to 5000 employees": "Medianas Empresas",
    "201 to 500 employees": "Pequeñas Empresas",
    "51 to 200 employees": "Pequeñas Grandes Empresas",
    "501 to 1000 employees": "Microempresas",
    "Unknown": "Empresas sin Información"
}
df['Size'] = df['Size'].replace(size_mapping)


## 7.Finalmente, genera una tabla pivote que muestre la media del salario estimado mínimo y la media del salario estimado máximo por tamaño de empresa

In [42]:
# Crear la tabla pivote
pivot_table_salary = pd.pivot_table(df, values=['Min Estimated Salary', 'Max Estimated Salary'], index='Size', aggfunc='mean')

# Mostrar la tabla pivote
pivot_table_salary

Max Estimated Salary  Min Estimated Salary
Size                                                                 
Empresas sin Información             110.500000             73.000000
Grandes Empresas                     138.875000             92.125000
Medianas Empresas                    137.461538             93.923077
Mega Empresas                        151.111111             97.888889
Microempresas                        146.235294            100.176471
Pequeñas Empresas                    141.142857             93.571429
Pequeñas Grandes Empresas            137.666667            100.666667